bluepyentity/register.py

```diff
@@ -0,0 +1,472 @@
+        return {"schema": schema, "ids": ids}
+
+    return None
+
+
+def _is_url(item):
+    return isinstance(item, str) and RE_URL.match(item)
+
+
+def _is_supported_type(type_):
+    # This is a bit dangerous as the type can be _Entity which would evaluate as True here
+    return get_entity_definition(type_) is not None
+
+
+def _wrap_linked_file_path(path):
+    """Wraps path as file URI if not already wrapped"""
+    return RE_NOT_FILE_URI.sub(FILE_URI, path)
+
+
+def _wrap_id(id_, type_):
+    """Wrap Nexus ID to an object"""
+    return {"type": type_, "id": id_}
+
+
+def _wrap_id_fetch(id_, forge):
+    """Find items with given IDs in Nexus and wrap them as objects."""
+    if _is_url(id_):
+        try:
+            return _wrap_id(id_, forge.retrieve(id_, cross_bucket=True).type)
+        except kgforge.core.commons.exceptions.RetrievalError as e:
+            raise RuntimeError(e) from e
+    elif isinstance(id_, list):
+        return [_wrap_id_fetch(i, forge) for i in id_]
+
+    return id_
+
+
+# NOTE: Should this work with bluepyentity.nexus.entityEntity?
+class Resource:
+    """Base class for resources to register."""
+
+    # This value should came from the schemas
+    required = set()
+
+    def __init__(self, definition, forge):
+        """Resource initializer.
+
+        Args:
+            definition (dict): attributes and their values to register
+            forge (kgforge.core.KnowledgeGraphForge): nexus-forge instance
+        """
+        # TODO: add registration to nexus.connector.Connector and use that instead of using
+        # kgforge methods directly
+        self._definition = definition
+        self._forge = forge
+
+        self._check_if_valid()
+
+        self._resource = self._create_resource()
+
+    @property
+    def definition(self):
+        """Adds defaults to the given definition and wraps the items in Nexus format."""
+        refined = self._format_attributes(self._definition)
+        refined = self._wrap_fields_with_ids(refined)
+
+        # Fields to not store in Nexus
+        refined.pop("brainRegion", None)
+
+        return self._with_defaults(refined)
+
+    @property
+    def possible_ids(self):
+        """Return the fields that possibly contains nexus ids"""
+        return self._schema_definition["ids"]
+
+    @property
+    def resource(self):
+        """Access the resource."""
+        return self._resource
+
+    @property
+    def schema(self):
+        """Schema for the entity."""
+        # TODO: figure out how to get this from Nexus and use it to validate types etc.
+        return self._schema_definition["schema"]
+
+    @property
+    def type(self):
+        """Return the Nexus type of the resource."""
+        return self._definition["type"]
+
+    @cached_property
+    def _schema_definition(self):
+        return get_entity_definition(self.type)
+
+    def register(self):
+        """Register the resource in Nexus."""
+        if existing := self._find_existing():
```

Comment:
If the resource isn't registered at this point, does this mean that it will search for all resources within search LIMIT and compare them with the current resource?
https://github.com/BlueBrain/nexus-forge/blob/master/kgforge/specializations/stores/bluebrain_nexus.py#L678
Even if the 100 first resources are not equal it doesn't mean that the resource doesn't exist.

Wouldn't it be preferable to override the `_find_existing()` in children classes to something more specific in addition to the resource type? For example, in the case of the DetailedCircuit if there is at least one circuit with the same circuitConfigPath then it is pretty certain we have encountered an existing resource.

For the base case, given that it's excessive to compare all the available resources in the bucket I would just register without checking at all. And sometimes you just want to clone a resource.

Reply:
See my comment above. I'll give this `class Resource` a do-over and let's see where we stand then when it is actually functioning with the pydantic stuff. :metal: